src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -657,9 +665,59 @@ internal static ulong GetSegmentSize()
             return RuntimeImports.RhGetGCSegmentSize();
         }
 
-        internal static T[] AllocateUninitializedArray<T>(int length)
+        internal static unsafe T[] AllocateUninitializedArray<T>(int length)
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>()
+#if !DEBUG
+            // Small arrays are allocated using `new[]` as that is generally faster.
+            //
+            // The threshold was derived from various simulations. 
+            // As it turned out the threshold depends on overal pattern of all allocations and is typically in 200-300 byte range.
+            // The gradient around the number is shallow (there is no perf cliff) and the exact value of the threshold does not matter a lot.
+            // So it is 256 bytes including array header.
+            || (Unsafe.SizeOf<T>() * length < 256 - 3 * IntPtr.Size)
+#endif
+            )
+            {
+                return new T[length];
+            }
+
+            var pEEType = EETypePtr.EETypePtrOf<T[]>();
+
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length), SR.ArgumentOutOfRange_NeedPosNum);
+
+            nuint nLength = (nuint)length;
+            nuint componentSize = pEEType.ComponentSize;
+            if (nLength > MaxArrayLength(componentSize))
+                throw new OutOfMemoryException();
+
+            nuint totalSize;
+            try
+            {
+                totalSize = checked(nLength * componentSize + pEEType.BaseSize);
+            }
+            catch (OverflowException)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            T[] array = null;
+            RuntimeImports.RhAllocateUninitializedArray(pEEType.RawValue, (UIntPtr)totalSize, (uint)length, Unsafe.AsPointer(ref array));
+
+            if (array == null)
+                throw new OutOfMemoryException();
+
+            return array;
+        }
+
+        private static nuint MaxArrayLength(nuint componentSize)
         {
-            return new T[length];
+            // Impose limits on maximum array length in each dimension to allow efficient 
+            // implementation of advanced range check elimination in future. We have to allow 
+            // higher limit for array of bytes (or one byte structs) for backward compatibility.
+            // Keep in sync with Array.MaxArrayLength in BCL.
+            return (componentSize == 1) ? (nuint)0X7FFFFFC7 : 0X7FEFFFFF;
```

Comment:
We could use actual `Array.MaxArrayLength`/`Array.MaxByteArrayLength` instead of a comment referring to them.

But I wonder whether it wouldn't be better to just move the length validation and size computation to native code - so that the `Rh` API doesn't have to specify both length and size. Native code already deals with computing sizes and validating limits around [here](https://github.com/dotnet/corert/blob/c42b7e691cd8ad75868773309d0f9cafe5c86956/src/Native/Runtime/gcrhenv.cpp#L251-L252) and [here](https://github.com/dotnet/corert/blob/4f4a7a04da4b59a1dd7dfcc019f0960c23108c97/src/Native/Runtime/portable.cpp#L128-L147) (this is how to do the math - it will end up calling the first helper for big arrays that then checks the limits).

Reply:
Fair point about MaxLength, i copied that without much thought from CoreCLR.
I put it into managed code because i cannot throw from unmanaged code (seeing the `if (array == null) throw` makes this a mood point but originally i was looking for the modified error message that is thrown for `ThrowOutOfMemoryDimensionsExceeded`; also i had the hunch that managed code is generally preferred). For the max_size part i had no clue on how to get my hands on `GCAllowVeryLargeObjects` on the C# side.

Reply:
Yes, managed code is preferred, but there's also preference for having related code together/sharing parts of it and the related code is on the native side. I don't have a strong opinion on that though.